internal/controller/pod_controller.go

```diff
@@ -46,16 +58,128 @@ type PodReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.0/pkg/reconcile
 func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	reqPod := &corev1.Pod{}
+	err := r.Get(ctx, req.NamespacedName, reqPod)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	annotations := reqPod.Annotations
+	if _, ok := annotations[constants.CatGateImagesHashAnnotation]; !ok {
+		logger.V(levelWarning).Info("pod annotation not found")
+		err := r.removeSchedulingGate(ctx, reqPod)
+		if err != nil {
+			logger.Error(err, "failed to remove scheduling gate")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	reqImagesHash := annotations[constants.CatGateImagesHashAnnotation]
+
+	pods := &corev1.PodList{}
+	err = r.List(ctx, pods, client.MatchingFields{constants.ImageHashAnnotationField: reqImagesHash}, client.InNamespace(req.Namespace))
+	if err != nil {
+		logger.Error(err, "failed to list pods")
+		return ctrl.Result{}, err
+	}
+
+	nodeSet := make(map[string]struct{})
+
+	numSchedulablePods := 0
+	numImagePulledPods := 0
+
+	for _, pod := range pods.Items {
+		if existsSchedulingGate(&pod) {
+			continue
+		}
+		numSchedulablePods += 1
+
+		allStarted := true
+		statuses := pod.Status.ContainerStatuses
+		for _, status := range statuses {
+			if status.State.Running == nil && status.State.Terminated == nil {
+				allStarted = false
+				break
+			}
+		}
 
-	// TODO(user): your logic here
+		if allStarted && len(pod.Spec.Containers) == len(statuses) {
+			nodeSet[pod.Spec.Hostname] = struct{}{}
+			numImagePulledPods += 1
+		}
+	}
 
-	return ctrl.Result{}, nil
+	capacity := len(nodeSet) * scaleRate
+
+	if capacity < minimumCapacity {
+		capacity = minimumCapacity
+	}
+
+	numImagePullingPods := numSchedulablePods - numImagePulledPods
+
+	if capacity > numImagePullingPods {
+		err := r.removeSchedulingGate(ctx, reqPod)
+		if err != nil {
+			logger.Error(err, "failed to remove scheduling gate")
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{
+		Requeue:      true,
+		RequeueAfter: 3 * time.Second,
```

Comment:
Specify either `requeue` or `requeueAfter`.
If `requeue` is specified, `ItemExponentialFailureRateLimiter` will be used.

Reply:
Will be using `requeue`